src/overlays/actors/ovl_En_Vali/z_en_vali.c

```diff
@@ -76,71 +82,732 @@ static ColliderCylinderInit D_80B288F0 = {
+}
+
+void EnVali_SetupRetaliate(EnVali* this) {
+    Animation_MorphToPlayOnce(&this->skelAnime, &gBariRetaliatingAnim, -5.0f);
+    Actor_SetColorFilter(&this->actor, 0x4000, 150, 0x2000, 30);
+    this->actor.params = BARI_TYPE_NORMAL;
+    this->bodyCollider.base.acFlags &= ~AC_ON;
+    this->actionFunc = EnVali_Retaliate;
+}
+
+void EnVali_SetupMoveArmsDown(EnVali* this) {
+    Animation_PlayOnce(&this->skelAnime, &gBariMovingArmsDownAnim);
+    this->actionFunc = EnVali_MoveArmsDown;
+}
+
+void EnVali_SetupBurnt(EnVali* this) {
+    this->timer = 2;
+    this->bodyCollider.base.acFlags &= ~AC_ON;
+    Actor_SetColorFilter(&this->actor, 0x4000, 150, 0x2000, 30);
+    this->actionFunc = EnVali_Burnt;
+}
+
+void EnVali_SetupDivideAndDie(EnVali* this, GlobalContext* globalCtx) {
+    s32 i;
+
+    for (i = 0; i < 3; i++) {
+        Actor_Spawn(&globalCtx->actorCtx, globalCtx, ACTOR_EN_BILI, this->actor.world.pos.x, this->actor.world.pos.y,
+                    this->actor.world.pos.z, 0, this->actor.world.rot.y, 0, 0);
+
+        this->actor.world.rot.y += 0x10000 / 3;
+    }
+
+    Item_DropCollectibleRandom(globalCtx, &this->actor, &this->actor.world.pos, 0x50);
+    this->timer = Rand_S16Offset(10, 10);
+    this->bodyCollider.base.acFlags &= ~AC_ON;
+    Audio_PlaySoundAtPosition(globalCtx, &this->actor.world.pos, 40, NA_SE_EN_BARI_SPLIT);
+    this->actor.flags &= ~1;
+    this->actor.draw = NULL;
+    this->actionFunc = EnVali_DivideAndDie;
+}
+
+void EnVali_SetupStunned(EnVali* this) {
+    Animation_MorphToPlayOnce(&this->skelAnime, &gBariWaitingAnim, 10.0f);
+    this->timer = 80;
+    this->actor.velocity.y = 0.0f;
+    Actor_SetColorFilter(&this->actor, 0, 255, 0x2000, 80);
+    this->bodyCollider.info.bumper.effect = 0;
+    Audio_PlayActorSound2(&this->actor, NA_SE_EN_GOMA_JR_FREEZE);
+    this->actor.velocity.y = 1.0f;
+    this->actionFunc = EnVali_Stunned;
+}
+
+void EnVali_SetupFrozen(EnVali* this) {
+    this->actor.velocity.y = 0.0f;
+    Actor_SetColorFilter(&this->actor, 0, 255, 0x2000, 36);
+    this->bodyCollider.base.acFlags &= ~AC_ON;
+    this->timer = 36;
+    this->actionFunc = EnVali_Frozen;
+}
+
+void EnVali_SetupReturnToLurk(EnVali* this) {
+    Animation_MorphToPlayOnce(&this->skelAnime, &gBariLurkingAnim, 10.0f);
+    this->actor.flags |= 0x10;
+    this->actor.flags &= ~1;
+    this->actionFunc = EnVali_ReturnToLurk;
+}
+
+void EnVali_DischargeLightning(EnVali* this, GlobalContext* globalCtx) {
+    static Color_RGBA8 effectPrimColor = { 255, 255, 255, 255 };
+    static Color_RGBA8 effectEnvColor = { 200, 255, 255, 255 };
+    Vec3f effectPos;
+    s32 i;
+    f32 cos;
+    f32 sin;
+    s16 effectYaw;
+
+    for (i = 0; i < 4; i++) {
+        cos = -Math_CosS(Camera_GetCamDirYaw(ACTIVE_CAM));
+        sin = Math_SinS(Camera_GetCamDirYaw(ACTIVE_CAM));
+        if (!((this->lightningTimer + (i << 1)) % 4)) {
+            effectYaw = (s16)(s32)Rand_CenteredFloat(12288.0f) + (i * 0x4000) + 0x2000;
```

Comment:
This s32 cast seems like it should be unnecessary?
```suggestion
            effectYaw = (s16)Rand_CenteredFloat(12288.0f) + (i * 0x4000) + 0x2000;
```